src/query/storage/promremote/storage.go

```diff
@@ -186,156 +189,234 @@ type promStorage struct {
+		Unit: queryOpt.Unit,
+		Tags: models.Tags{
+			Opts: queryOpt.Tags.Opts,
+		},
+		FromIngestor: queryOpt.FromIngestor,
+	}
+	cp.Datapoints = make([]ts.Datapoint, 0, len(queryOpt.Datapoints))
+	cp.Datapoints = append(cp.Datapoints, queryOpt.Datapoints...)
+
+	cp.Tags.Tags = make([]models.Tag, 0, len(queryOpt.Tags.Tags))
+	cp.Tags.Tags = append(cp.Tags.Tags, queryOpt.Tags.Tags...)
+	/*
+	// In case deeper copying is needed
+	for i, tag := range queryOpt.Tags.Tags {
+		tagCopy := models.Tag{
+			Name:  make([]byte, len(tag.Name)),
+			Value:  make([]byte, len(tag.Value)),
+		}
+		copy(tagCopy.Name, tag.Name)
+		copy(tagCopy.Value, tag.Value)
+		cp.Tags.Tags[i] = tagCopy
+	}
+	*/
+	return cp
+}
+
 func (p *promStorage) Write(ctx context.Context, query *storage.WriteQuery) error {
-	if query != nil {
-		p.queryQueue <- query
-		p.enqueued.Inc(1)
+	if query == nil {
+		return nil
 	}
+	if query.Options().DuplicateWrite {
+		// M3 call site may write the same data according to different storage policies.
+		// See downsampleAndWriter in src/cmd/services/m3coordinator/ingest/write.go
+		p.dupWrites.Inc(1)
+		return nil
+	}
+	if query.Options().FromIngestor {
+		// src/cmd/services/m3coordinator/ingest/m3msg/ingest.go reuses a WriteQuery object to write different
+		// time series by calling ResetWriteQuery(). We need to make a copy of the WriteQuery object to avoid
+		// race conditions.
+		p.ingestorWrites.Inc(1)
+		queryCopy, err := storage.NewWriteQuery(deepCopy(query.Options()))
+		if err != nil {
+			p.enqueueErr.Inc(1)
+			p.logger.Error("error copying write", zap.Error(err),
+				zap.String("write", query.String()))
+			return err
+		}
+		query = queryCopy
+	}
+	p.queryQueue <- query
+	p.enqueued.Inc(1)
 	return nil
 }
 
-func (p *promStorage) writeBatch(ctx context.Context, tenant tenant, queries []*storage.WriteQuery) error {
+func (p *promStorage) writeBatch(ctx context.Context, tenant tenantKey, queries []*storage.WriteQuery) error {
 	logSampling := rand.Float32()
 	if logSampling < logSamplingRate {
 		p.logger.Debug("async write batch",
-			zap.String("tenant", tenant.name),
-			zap.Duration("retention", tenant.attr.Retention),
-			zap.Duration("resolution", tenant.attr.Resolution),
+			zap.String("tenant", string(tenant)),
 			zap.Int("size", len(queries)))
 	}
+	// TODO: remove this double check once the bug is confirmed to be fixed.
+	{
+		correctWq := make([]*storage.WriteQuery, 0, len(queries))
+		wrongTenants := 0
+		for _, wq := range queries {
+			correctTenant := p.getTenant(wq)
+			if correctTenant == tenant {
+				correctWq = append(correctWq, wq)
+			} else {
+				wrongTenants++
+				if rand.Float32() < 0.01 {
+					p.logger.Error("dropping a write because of a wrong tenant",
+						zap.String("expected_tenant", string(correctTenant)),
+						zap.String("actual_tenant", string(tenant)),
+						zap.String("time_series", wq.String()),
+						zap.Bool("from_ingestor", wq.Options().FromIngestor),
```

Comment:
shall we count how many wrong tenants are from pre-agg and if there are any from raw in metrics?

Reply:
Not very important. If this counter ever gets bumped later on, we'd have to dive deep into it.